### PR TITLE
fix .netrc authentication

### LINKED ISSUE #2480: netrc x-oauth-basic no longer works since 1.2.0-rc1
- Gitea version: 1.2.0-rc1, 1.2.0-rc2
- Git version: 2.11.0
- Operating system: docker image gitea/gitea on Debian 4.9.30-2+deb9u3 (kernel 4.9.0-3-amd64)
- Database:
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL)
  - [ ] No
  - [ ] Not relevant
- Log gist:

## Description
Hello, coming here from [drone's discourse](https://discourse.drone.io/t/cannot-clone-private-gitea-repo-could-not-read-username-for-0-8-0-rc-3-gitea-1-2-0-rc2/624/3) forum, it is no longer possible to authenticate using the x-oauth-basic as used previously in version 1.1.4

`git clone https://{TOKEN}:x-oauth-basic@gitea.my.domain/Org/Repo.git`

yields 

`remote: invalid credentials
fatal: Authentication failed for 'https://{TOKEN}:x-oauth-basic@gitea.my.domain/Org/Repo.git/'`

In version 1.1.4 it succesfully clones the repository.

Using

`git clone https://{USERNAME}:{TOKEN}@gitea.my.domain/Org/Repo.git`

it is possible to clone the repository in version 1.2.0-rc1 and 1.2.0-rc2

This unfortunatelly breaks the drone builds as the [plugins/git](https://github.com/drone-plugins/drone-git) docker image uses the x-oauth-basic to clone the repository. I am not sure if there was a decistion to drop the x-oauth-basic or if it is a bug since I did not find a relevant issue or a mention of the change in changelog. So it is not clear whether the issue should be fixed from the side of gitea or if the drone plugin should be updated to support the current scheme.

## How to reproduce:
1. Create a new private repository on try.gitea.com
2. Add a new application under your profile settings
3. Try to clone the repository using the application's token `git clone https://{OAUTH_TOKEN_REDACTED}:x-oauth-basic@try.gitea.com/Org/Repo.git`
4. Verify the token is valid by cloning the repo using your username `git clone https://{USERNAME}:{TOKEN}@try.gitea.com/Org/Repo.git`

### PR DESCRIPTION
**Pull request by daviian**
Fixes #2480

Reverted a part of #2184 to allow token usage without the need of 2FA.
But token is required if 2FA is enabled!

### CODE REVIEW

**Comment by lunny** [comment]
@daviian but since https://github.com/drone/drone/pull/2241 merged. This PR will broke it again?

**Review by lafriks — CHANGES_REQUESTED** [review]
Yes I think that would break fixed drone

**Comment by daviian** [comment]
And current drone only works with latest gitea release... and that's not good at all. Every gitea user that uses drone is now forced to upgrade

**Comment by lafriks** [comment]
Drone support only 1.2 anyway as only 1.2 has status api

**Comment by daviian** [comment]
But as you can see in issues users use drone in 1.1.4 and earlier as well.

**Comment by daviian** [comment]
@lunny @lafriks I changed it to support both options. Username as token as well as password as token.

So no break in drone and backwards compatibility.

**Review by lafriks — APPROVED** [review]

**Comment by lunny** [comment]
@daviian gitea 1.1.4 could use Gogs driver of drone.

**Comment by lafriks** [comment]
@daviian I still think it's not a good idea to do this as .netrc anyway supports both username&password and drone is only that use this and now that it has already changed to use new way, let's than keep just one way not to have problems later.

**Comment by lunny** [comment]
@lafriks but that ask people upgrade their drone to latest. So I will give this LGTM

File: routers/repo/http.go [hunk 1]
```diff
@@ -161,16 +171,22 @@ func HTTP(ctx *context.Context) {
...
 
-				if authUser.ID != token.UID {
-					ctx.HandleText(http.StatusUnauthorized, "invalid credentials")
-					return
+				if isUsernameToken {
+					authUser, err = models.GetUserByID(token.UID)
+					if err != nil {
+						ctx.Handle(http.StatusInternalServerError, "GetUserByID", err)
+					}
+				} else {
```
Comment (Morlinest) [inline routers/repo/http.go]: `else if`

**Comment by lafriks** [comment]
LGTM

File: routers/repo/http.go [hunk 2]
```diff
@@ -161,16 +171,22 @@ func HTTP(ctx *context.Context) {
 					return
 				}
 
-				if authUser.ID != token.UID {
-					ctx.HandleText(http.StatusUnauthorized, "invalid credentials")
-					return
+				if isUsernameToken {
+					authUser, err = models.GetUserByID(token.UID)
+					if err != nil {
+						ctx.Handle(http.StatusInternalServerError, "GetUserByID", err)
```
Comment (Morlinest) [inline routers/repo/http.go]: missing `return`?

**Review by Morlinest — CHANGES_REQUESTED** [review]

File: routers/repo/http.go [hunk 3]
```diff
@@ -139,19 +139,29 @@ func HTTP(ctx *context.Context) {
...
-						ctx.HandleText(http.StatusUnauthorized, "invalid credentials")
-					} else {
-						ctx.Handle(http.StatusInternalServerError, "GetUserByName", err)
+				var authToken string
+				isUsernameToken := len(authPasswd) == 0 || authPasswd == "x-oauth-basic"
+
+				if isUsernameToken {
+					// Assume username is token
+					authToken = authUsername
+				} else {
```
Comment (Morlinest) [inline routers/repo/http.go]:
```
// Assume username is token
authToken := authUsername

if !isUsernameToken {
	// Assume password is token
	authToken = authPasswd
```

**Review by Morlinest — CHANGES_REQUESTED** [review]

**Comment by daviian** [comment]
Make LG-TM work

**Review by Morlinest — APPROVED** [review]

**Comment by lafriks** [comment]
Make it work again

**Comment by lafriks** [comment]
@daviian please backport to 1.2 branch

**Pull request merged**